### ISSUE TITLE
Add syntax highlighting in Regex mode

### ISSUE THREAD

**Issue by adamfranco** [atom/find-and-replace#458]
Highlighting syntax in Regex mode would be a nice usability enhancement. This is something the BBEdit editor does very well, highlighting regular expressing special characters in one color and escaped characters in another:

![bbedit-find-highlight-a](https://cloud.githubusercontent.com/assets/25242/8853776/5c8c700a-312a-11e5-9603-536a3f81eafa.png)

Doing this makes it much easier to construct and modify regex search strings as code often contains  many characters that are also special regular-expression syntax characters. Additionally, highlighting placeholders and escaped characters in the replacement string would make them stand out more from the static replacement text.

**Comment by adamfranco** [comment]
I'd be happy to work on a pull request for this enhancement, but I can imagine several different implementation scenarios (such as building a stand-alone regular-expression syntax highlighter and then leveraging that, versus adding highlighting right into the find-field handling code directly). Any suggestions on how best to implement this would be welcome.

**Comment by benogle** [comment]
You might be able to write a [grammar](https://github.com/atom/language-javascript) for regular expressions then use `setGrammar()` on the inputs (mini editors). Though I'm not totally sure how mini editors handle grammars. I would try 2 things:

* Set a grammar on a mini editor. Does it work?
* Write a grammar for regex, or find a textmate grammar and convert it with `apm`.

**Comment by adamfranco** [comment]
I tried adding the source.js grammar to the editor in the toggleRegexOption function as a test and found that the mini-editor would show syntax highlighting:

      toggleRegexOption: =>
         @updateModel {pattern: @findEditor.getText(), useRegex: not @findModel.useRegex}
         @selectFirstMarkerAfterCursor()
         @updateOptionsLabel()
    +    grammar = atom.grammars.grammarForScopeName('source.js')
    +    @findEditor.getModel().setGrammar(grammar)

Result:
![screen shot 2015-07-28 at 6 01 06 pm](https://cloud.githubusercontent.com/assets/25242/8945108/abb39dda-3552-11e5-8c1b-377973aa431f.png)

Next task: finding or creating a regular-expression grammar...

**Comment by benogle** [comment]
Cool!

**Comment by adamfranco** [comment]
I've now gotten this partially working via 766968a. Below are som other to-dos that need to be tackled before I consider this enhancement complete:

* Make the any-character wild-card `.` stand out more. Currently I have ranges and character classes highlighted with green, indicating that they will match a set of characters rather than a single explicit character. Unfortunately, the green doesn't stand out much from a black `.` when buried in an expression. Maybe swapping the green for a darker red than the keywords would be good option.
* Add highlighting of replacement placeholders.
* Add highlighting in the find-in-project mini-editor.

**Comment by adamfranco** [comment]
...and here's what the find dialog looks like with 766968a in place:

<img width="955" alt="screen shot 2015-07-30 at 1 49 00 am" src="https://cloud.githubusercontent.com/assets/25242/8976573/3106f238-365d-11e5-92f9-295d30100b19.png">

**Comment by benogle** [comment]
Looks like it's coming along!

**Comment by adamfranco** [comment]
I'm running my [regex_syntax_highlighting branch](https://github.com/adamfranco/find-and-replace/tree/regex_syntax_highlighting) during my day-job and finding that the enhancement is working pretty well for my purposes so far. Still to do:

* Move the highlight styling to the themes as you suggested. I imagine this will also make it easier for end-users to tweak the styles to their liking.
* Create a grammar for replacement placeholders.
** Add support for the replacement placeholder grammar to find-and-replace.
* Wait on a resolution to atom/language-python#83 for improved styling of wildcard characters, though this isn't really a blocker for this find-and-replace issue since it will only affect the theming.

Thanks again for your feedback, the guidance has been very helpful!

**Comment by benogle** [comment]
:+1: 

Are there any differences between the py regular expressions and js regexes that could be lead to strange highlighting? Wondering if a totally separate regex syntax could be a good thing. Though it's awesome that you found one that is already bundled. I had no idea it existed.

**Comment by adamfranco** [comment]
Hmm, I found the python one first and just ran with it, but I just tried switching to the `source.js.regexp` grammar which also seems to be bundled with the default install and it seems to work fine. If the JS syntax is what is actually being used for searching, I might as well use its grammar.

**Comment by adamfranco** [comment]
I've now moved syntax highlighting to the themes on the following branches:
* https://github.com/adamfranco/atom-dark-ui/tree/highlight_find-and-replace_regexp
  ![atom-dark](https://cloud.githubusercontent.com/assets/25242/9048302/240fa6ea-3a06-11e5-9b85-003da005e586.png)
  
* https://github.com/adamfranco/atom-light-ui/tree/highlight_find-and-replace_regexp
  ![atom-light](https://cloud.githubusercontent.com/assets/25242/9048310/31a63bf2-3a06-11e5-89e1-a688ed29cb14.png)
  
* https://github.com/adamfranco/one-dark-ui/tree/highlight_find-and-replace_regexp
  ![one-dark](https://cloud.githubusercontent.com/assets/25242/9048314/3a99c7d8-3a06-11e5-9a69-306c48b10dda.png)
  
* https://github.com/adamfranco/one-light-ui/tree/highlight_find-and-replace_regexp
  ![one-light](https://cloud.githubusercontent.com/assets/25242/9048317/41a9f4bc-3a06-11e5-987b-cdbc26ea2e8b.png)
  
Still to do:
* Add a grammar and theming for replacements.
* Add wildcard differentiation to the `source.js.regexp` grammar.

**Comment by benogle** [comment]
I'm not a fan of the method for adding the coloring: it adds several variables, has find and replace specific bits, and the colors are hardcoded into the theme, so might differ from the syntax theme. 

Ideally this coloring is in the syntax themes in a pretty generic way. The only problem with that is the user might have a dark syntax theme with a light UI theme, and the mini-editor colors are currently determined by the UI theme. 

@simurai, thoughts on this? What do you think we should do? Maybe these go in the syntax themes, and the people with a mismatched syntax/ui theme might have slightly washed out coloring?

**Comment by simurai** [comment]
It would be nice if the syntax styling could be done only from the find-and-replace package. Just so that all themes would get it automatically (they could still override it, if they really want).

I think there are 3 colors needed in total. Maybe we could use [these 3](https://github.com/atom/atom/blob/master/static/variables/ui-variables.less#L11-L13) from `ui-variables.less`. It is a bit a stretch and not really intended like that, but would make sure that the contrast works in most themes.

```less
// in find-and-replace/styles/find-and-replace-syntax.less or so

@import "ui-variables";

@find-and-replace-regexp-color-escape: @text-color-warning;
@find-and-replace-regexp-color-range: @text-color-info;
@find-and-replace-regexp-color-character-class: @text-color-info;
@find-and-replace-regexp-color-set: inherit;
@find-and-replace-regexp-color-wildcard: @text-color-info;
@find-and-replace-regexp-color-keyword: @text-color-success;
@find-and-replace-regexp-color-punctuation: @text-color-success;
@find-and-replace-regexp-weight-escape: normal;
@find-and-replace-regexp-weight-range: normal;
@find-and-replace-regexp-weight-character-class: normal;
@find-and-replace-regexp-weight-set: inherit;
@find-and-replace-regexp-weight-wildcard: bold;
@find-and-replace-regexp-weight-keyword: normal;
@find-and-replace-regexp-weight-punctuation: normal;

.find-container atom-text-editor[mini]::shadow {
  .regexp {
    .escape {
      color: @find-and-replace-regexp-color-escape;
      font-weight: @find-and-replace-regexp-weight-escape;
    }
    .range{
      color: @find-and-replace-regexp-color-range;
      font-weight: @find-and-replace-regexp-weight-range;
    }
    .character-class {
      color: @find-and-replace-regexp-color-character-class;
      font-weight: @find-and-replace-regexp-weight-character-class;
    }
    .set {
      color: @find-and-replace-regexp-color-set;
      font-weight: @find-and-replace-regexp-weight-set;
    }
    .punctuation {
      color: @find-and-replace-regexp-color-punctuation;
      font-weight: @find-and-replace-regexp-weight-punctuation;
    }
    .keyword {
      color: @find-and-replace-regexp-color-keyword;
      font-weight: @find-and-replace-regexp-weight-keyword;
    }
  }
}

```

ps. The variables could still be re-ordered. Just added them without actual testing.

**Comment by adamfranco** [comment]
Thanks for the feedback, @simurai. Given that there are only 3 colors in use, would it be preferable to condense the styles like this to concisely capture the intended variance in style:

    // in find-and-replace/styles/find-and-replace-syntax.less or so

    @import "ui-variables";

    @find-and-replace-regexp-color-escape: @text-color-warning;
    @find-and-replace-regexp-color-range: @text-color-info;
    @find-and-replace-regexp-color-keyword: @text-color-success;

    .find-container atom-text-editor::shadow {
      // Styles for regular expression highlighting
      .regexp {
        .escape {
          color: @find-and-replace-regexp-weight-escape;
        }
        .range, .character-class, .wildcard {
          color: @find-and-replace-regexp-color-range;
        }
        .wildcard {
          font-weight: bold;
        }
        .set {
          color: inherit;
        }
        .keyword, .punctuation {
          color: @find-and-replace-regexp-color-keyword;
          font-weight: normal;
        }
      }
    }

or leave it more verbose with more variables for themes to override if desired? Having not really worked with Atom themes I'm unclear were to draw the line on verbosity/overriding-via-variables.

If I understand correctly, going the condensed route in the package's styles would allow a theme to optionally override the color scheme by only adding three lines like this, correct?

    @find-and-replace-regexp-color-escape: #96CBFE;
    @find-and-replace-regexp-color-range: #99CC99;
    @find-and-replace-regexp-color-keyword: #FF3333;

**Comment by simurai** [comment]
> would it be preferable to condense the styles like this to concisely capture the intended variance in style

Yes, that seems much simpler. :+1: 

> If I understand correctly, going the condensed route in the package's styles would allow a theme to optionally override the color scheme by only adding three lines like this, correct?

No, I think that would have no effect because of the way the theme and package files get loaded. Themes that want to override it could:

1. change the `@text-color-warning; @text-color-info; @text-color-success;`.
2. If those text colors should remain the same, they would have to override it using normal CSS selectors. Basically use the same as the find-and-replace package:

```less
.find-container atom-text-editor::shadow {
  // Styles for regular expression highlighting
  .regexp {
    .escape {
      color: #my-color-1;
    }
    .range, .character-class, .wildcard {
      color: #my-color-2;
    }
    .keyword, .punctuation {
      color: #my-color-3;
    }
  }
}
```

**Comment by adamfranco** [comment]
With 2a73556, highlighting now has sensible defaults in all of the themes below, without any hard-coded coloring in the package or extra variables that don't accomplish anything:

* ![one-light](https://cloud.githubusercontent.com/assets/25242/9067773/7d07726c-3aad-11e5-8f0e-7c3c0a3337d7.png)
* ![one-dark](https://cloud.githubusercontent.com/assets/25242/9067774/7d08c4aa-3aad-11e5-8422-b26bac8923f7.png)
* ![atom-light](https://cloud.githubusercontent.com/assets/25242/9067776/7d0daa2e-3aad-11e5-916b-b1ed4e1f8b32.png)
* atom-dark:![atom-dark](https://cloud.githubusercontent.com/assets/25242/9067772/7d05b332-3aad-11e5-8f35-45566edc41f1.png)
* unity:
  ![unity](https://cloud.githubusercontent.com/assets/25242/9067775/7d091be4-3aad-11e5-89b4-21ae83895c9b.png)

**Comment by simurai** [comment]
:+1: Great.

The red in atom-dark looks a bit dark, but that should be tweaked in the theme. Or have you tried using `@text-color-warning` instead of `@text-color-error`, that might look a bit brighter.

**Comment by adamfranco** [comment]
Here are screen-grabs with `@text-color-warning`.  I personally find the orange a bit less pleasing than the red of `@text-color-error` and the red is a little more in line with my original conception of the feature, but I'm open to input.

* one-light:
  ![one-light](https://cloud.githubusercontent.com/assets/25242/9087264/83b0fa34-3b57-11e5-9cdb-9adead93fdd7.png)

* one-dark:
  ![one-dark](https://cloud.githubusercontent.com/assets/25242/9087262/83ae031a-3b57-11e5-930a-7ad720c22170.png)
  
* atom-light:
  ![atom-light](https://cloud.githubusercontent.com/assets/25242/9087261/83aca358-3b57-11e5-94c9-6b69967d64f8.png)
  
* atom-dark:
  ![atom-dark](https://cloud.githubusercontent.com/assets/25242/9087265/83b34d70-3b57-11e5-811e-fb3d7b1d70c8.png)
  
* unity
  ![unity](https://cloud.githubusercontent.com/assets/25242/9087263/83af29de-3b57-11e5-954b-93d0b29595cb.png)

**Comment by benogle** [comment]
@adamfranco you should open a PR with these changes

**Comment by adamfranco** [comment]
Will do. I'll open a new issue to track work on adding highlighting for the replacement placeholders since that enhancement will have a few more moving parts (a new grammar as well as tweaks to the find-and-replace package) and this current work is usable without that extra feature.

**Comment by simurai** [comment]
Yeah, `@text-color-warning` looks a bit weaker in light themes. But somehow I still tend to prefer `warning`, just because `error` might suggest something is invalid or so. But my opinion is not too strong about it. I'm ok leaving it with `error` if you prefer that. It's already a big improvement either way.
